tests/ui/lint/unused_parens_cast.rs

```diff
@@ -0,0 +1,101 @@
+// check-pass
+
+#![warn(unused_parens)]
+
+struct Foo(f32);
+
+impl Foo {
+    pub fn f32(self) -> f32 {
+        64.0f32
+    }
+}
+
+fn bar() -> f32 {
+    3.0f32
+}
+
+mod inner {
+    pub mod yet_inner {
+        pub mod most_inner {
+            pub static VERY_LONG_PATH: f32 = 99.0f32;
+        }
+    }
+}
+
+fn basic_test() {
+    // should fire
+    let one = 1.0f32;
+    let _ = (one) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    let _ = (inner::yet_inner::most_inner::VERY_LONG_PATH) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    let _ = (Foo(1.0f32).0) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    let _ = (Foo(1.0f32).f32()) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    let _ = (bar()) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    let baz = [4.0f32];
+    let _ = (baz[0]) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    // following is technically unnecessary, but is allowed because it may confusing.
+    let _ = (-1.0f32) as f64;
+    let x = Box::new(-1.0f32);
+    let _ = (*x) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    // cast is left-assoc
+    let _ = (true as u8) as u16;
+    //~^ WARN unnecessary parentheses around cast expression
+    // should not fire
```

Comment:
Please add unary `-` to this block (and other unary operator as well?)

Reply:
Does your intention include `*expr as Ty` and `&expr as Ty`?